### PR TITLE
Move secrets to Key Vault

### PR DESCRIPTION
**Pull request by senthilkungumaraj**
Moved secrets to Key Vault. Includes:

1. Creation of missing secrets in Key Vault through bicep templates
2. Changes to AKS resource in bicep template
3. Using csi-secrets-store-identity-access to access secrets in Key Vault and expose them as env vars in pod
4. Updates to the deployment scripts

### CODE REVIEW

File: deploy/scripts/install.sh
Comment (brianloss) [inline deploy/scripts/install.sh:46]: Is Workload Identity available beyond the commercial cloud yet? E.g., sovereign clouds? Pod Identity is an option if not--it's definitely deprecated, though, and supposed to be replaced with workload identity so I wouldn't change/add support unless we're missing workload identity support.

File: deploy/scripts/configure.sh
Comment (brianloss) [inline deploy/scripts/configure.sh:70]: This is pretty new and might require an az client update. Not sure if we need to document that somewhere or update instructions. I like the approach, though. (Also see my other comment about workload identity.)

File: deploy/bicep/modules/storage.credentials.to.keyvault.bicep
Comment (brianloss) [inline deploy/bicep/modules/storage.credentials.to.keyvault.bicep:36]: Is it possible to also retrieve this from the storageAccount resource so that we're not constructing the string? It would just make it a little less fragile in case the format of this string changes...

File: deploy/bicep/modules/appinsights.credentials.to.keyvault.bicep
Comment (brianloss) [inline deploy/bicep/modules/appinsights.credentials.to.keyvault.bicep:16]: If the template were to be re-applied, we'd create another deployment here. That might be ok--if so then doesn't it also need to be done for appinsightsSecrets in main.bicep above? Seems like we want the deployment to be idempotent if possible, though so using a fixed name would be better if it's possible to do so.

File: deploy/bicep/groups/monitoring.bicep
Comment (brianloss) [inline deploy/bicep/groups/monitoring.bicep:12]: These don't appear to be used, or I missed it if they are. Can these be removed if they're not needed?

**Review by brianloss — COMMENTED** [review]
Looking pretty good! A few small comments...

File: deploy/bicep/groups/monitoring.bicep
Comment (senthilkungumaraj) [reply deploy/bicep/groups/monitoring.bicep]: Removed the unused vars.

File: deploy/bicep/modules/appinsights.credentials.to.keyvault.bicep
Comment (senthilkungumaraj) [reply deploy/bicep/modules/appinsights.credentials.to.keyvault.bicep]: Removed the utcValue and instead added uniqueString() with subscription id.

File: deploy/scripts/install.sh
Comment (senthilkungumaraj) [reply deploy/scripts/install.sh:46]: Good catch. I will try to get an answer to this before merging this PR.

File: deploy/scripts/configure.sh
Comment (senthilkungumaraj) [reply deploy/scripts/configure.sh:70]: Updated the README.md file under deploy to include the additional pre-requisites including the az cli version (2.40.0 or later) to support `az identity federated-credential create` command.

File: deploy/bicep/modules/storage.credentials.to.keyvault.bicep
Comment (brianloss) [reply deploy/bicep/modules/storage.credentials.to.keyvault.bicep:36]: Yeah, that's too bad. I looked around as well and couldn't find anything. Given that, what you have is the best approach.

**Review by brianloss — APPROVED** [review]
Approved, but be sure to update if you find that workload identity isn't "ready enough". If that were the case, it might be good to keep it as an option but also offer another connection method. There is more than just pod identity. For example, you can use the kubelet identity (though that means anyone who can create pods on the cluster could access the storage account). That might be ok as a fallback. Since we are creating the k8s cluster it's safe to assume single tenancy, but if someone brought their own they might not want that approach. Long-term workload identity is definitely the right solution.

File: deploy/scripts/install.sh
Comment (senthilkungumaraj) [reply deploy/scripts/install.sh:46]: Worked with @karthick-rn today to validate if workload identity is available in Fairfax. We went through creating a new AKS Cluster and configuring the Managed Identity and Service Account. Then federated the Svc Account with the Managed Identity. Looks like the workload identity is available in Fairfax. 

One finding from the exercise was that the Kubernetes version 1.25.4 is not available. The latest available version is 1.25.2. I will change the version in this PR.

Comment (brianloss) [reply deploy/scripts/install.sh:46]: Good news about workload identity (and federated credentials for managed identities) in sovereign clouds.

Chasing k8s versions can be tough since they release frequently and also deprecate frequently. Setting the k8s version to an empty string in the templates should cause it to use the default version defined for the particular cloud and region. I think that might be the way to go--if you want to choose a specific version, then you can do so by setting the k8s version explicitly in your template.

**Review by waystilos — APPROVED** [review]

**Review by mandarinamdar — APPROVED** [review]
Thanks for addressing the comments, approving.

File: deploy/scripts/install.sh
Comment (senthilkungumaraj) [reply deploy/scripts/install.sh:47]: As per our discussion, I have set the Minimum AKS version to 1.22.0 (Workload Identity supported version). Added logic in the install.sh to get the latest version and error out if the version is less than 1.22.0.

File: deploy/scripts/install.sh
Comment (brianloss) [inline deploy/scripts/install.sh:52]: No need for `jq`. The orchestratorVersion can be pulled using an az cli query.
```suggestion
AKS_VERSION=$(az aks get-versions -l $LOCATION --query "orchestrators[].orchestratorVersion" -otsv | sort -rV | head -n1)
```
Thinking about it more, you may also not want to take the latest version since that's typically a preview version. Maybe you want the latest non-preview version from the list instead:
```suggestion
AKS_VERSION=$(az aks get-versions -l $LOCATION --query 'orchestrators[?!isPreview].orchestratorVersion' -otsv | sort -rV | head -n1)
```
I haven't verified, but I believe this is the version chosen if you just don't pass kubernetesVersion to the bicep `Microsoft.ContainerService/managedClusters` resource.

File: deploy/scripts/install.sh
Comment (brianloss) [inline deploy/scripts/install.sh:54]: AKS has an ["N-2" version policy](https://learn.microsoft.com/en-us/azure/architecture/operator-guides/aks/aks-upgrade-practices) and the [latest version](https://kubernetes.io/docs/home/supported-doc-versions/#version-latest) is 1.26, which means no Azure cluster should have 1.22 available and this if check will never succeed. It should also be impossible for a cluster to be created with a version older than 1.22, which means this whole block is probably unnecessary.

**Review by waystilos — APPROVED** [review]

**Review by waystilos — APPROVED** [review]

File: deploy/scripts/install.sh
Comment (senthilkungumaraj) [reply deploy/scripts/install.sh]: Added the suggested change to the latest commit.

File: deploy/scripts/install.sh
Comment (senthilkungumaraj) [reply deploy/scripts/install.sh]: Added the suggested change to the latest commit.

**Pull request merged**
